DESIGN.md

```diff
@@ -0,0 +1,362 @@
+SymEval Term`. One can then think of this as a function of type:
+
+```haskell
+SymEvalEnv -> (Term, SymEvalSt) -> [(Term, SymEvalSt)]
+```
+
+Such a function takes a term in the input language and a symbolic evaluation
+state. It tries to apply one step of beta-reduction to the term, which may yield
+zero, one or several potential terms and updated states.
+
+As an example, let us assume that our input term is the body of the `foldr`
+function in [the example language], specialised for lists of `Integer`s (for
+readability, we remove the type applications).
+
+```haskell
+match_List l
+  e
+  (\(x : Integer) (xs : List Integer) . f x (foldr f e xs))
+```
+
+and that our state specifies that:
+
+- `e` is an integer, without giving its value,
+- `f` is the function `\(x : Integer) (y : Integer) . x + y`, and
+- `l` is a list of integers, without giving its value.
+
+One step of the symbolic engine would then have to explore both branches of the
+pattern matching, and therefore it would return two pairs `Term, SymEvalSt`:
+
+1. In the first pair, the `Term` would be `e` itself and the state would be
+   updated to account for the fact that we now know that `l = Nil`.
+2. In the second pair, the `Term` would be `f x (foldr f e xs)` and the state
+   would be updated to account for the fact the we now have two more variables,
+   `x` and `xs`, that are an integer and a list of integers respectively, and an
+   additional constraint, `l = Cons x xs`.
+
+The whole symbolic execution consists in running this one step over and over
+again until it is not possible to reduce the terms anymore. In the example
+above, our first term/state is now blocked because there isn't much one can do
+with an integer. The second term/state is not blocked because one can replace
+`f` by its definition and continue unfolding from there.
+
+#### The symbolic evaluation state — `SymEvalSt`
+
+Defined in [`Pirouette.Symbolic.Eval.Types`], the symbolic state basically
+contains the following:
+
+1. A context of variables, that is a map from variable names to their types.
+   This context is typically called Γ in literature, and `gamma` in the code.
+
+2. A set of constraints linking variables together, for instance `l == Cons x
+   xs`, `k != Nil` or `x = 7`. It is in fact not a set, but a complex
+   union-find-based type.
+
+3. Statistics on the execution, for instance the number of unfoldings that took
+   place. Despite the name, those statistics can have an important impact on the
+   behaviour of the engine because they are observed to decide whether to stop
+   the execution or not.
+
+as well as some extra fields necessary for administrative reasons. Those field
+names are prefixed by `sest` in the code.
+
+#### The symbolic evaluation environment — `SymEvalEnv`
+
+Defined in [`Pirouette.Symbolic.Eval`], the symbolic environment contains the
+following:
+
+1. A set of unordered definitions in which to evaluate the term. For instance,
```

Comment:
How is this conceptually different from the constraints that link variables to variables and values in `SymEvalSt`?

Reply:
I don't think it is conceptually different. Constraints are more expressive, and I think we could easily encode the set of definitions into constraints. I even think that's what we must do to send the set of definitions to the solver, but I am unsure of this, and there might be performance considerations at play here. It might be worth investigating, though, because that could potentially lead to a simplification of the code. What do you want to do with this for the design document?

Reply:
I would put a note in the documentation to say as much. It is fair to defer diving further into the matter.

Reply:
Done in fbe4e22.

Reply:
And I opened https://github.com/tweag/pirouette/issues/188.